### ISSUE TITLE
linkerd doesn't reconnect to server after server restart for h2 protocol

### ISSUE THREAD

**Issue by irachex** [BuoyantIO/linkerd#758]
Hi, I'm very exciting about h2 protocol support. But we run into an issue trying h2.
### Problem

linkerd doesn't reconnect to server after server restart
### Steps to reproduce
1. Start a h2 server
   I use python `h2` to write a simple HTTP/2 server [source](https://gist.github.com/irachex/e7471d0a9b9d0e4bc0dd8a2cbc6557eb)
   
   ``` python
   python h2server.py 10280
   ```
2. Start linkerd 0.8.2 with follow config
   
   ``` yaml
   routers:
   - protocol: h2
     experimental: true
     baseDtab: |
        /h2=> /$/inet/127.0.0.1/10280;
     servers:
     - port: 10281
       ip: 0.0.0.0
   ```
3. Send request to linkerd and it works well
   
   ```
   curl --http2 127.0.0.1:10281/test
   ```
4. Stop server and restart it. Send request again. And linkerd report `com.twitter.finagle.DroppedWriteException` for all later requests.
   
   ```
   E 1019 03:39:50.858 THREAD69: demuxer
   com.twitter.finagle.ChannelClosedException: ChannelException at remote address: /127.0.0.1:10280. Remote Info: Not Available
   
   E 1019 03:39:56.310 THREAD75 TraceId:fe08e43521a5051c: Exception propagated to the root monitor!
   com.twitter.finagle.DroppedWriteException
   at com.twitter.finagle.netty4.transport.ChannelTransport.write(ChannelTransport.scala:100)
   at com.twitter.finagle.buoyant.h2.netty4.Netty4H2Writer$$anon$1.write(Netty4H2Writer.scala:75)
   at com.twitter.finagle.buoyant.h2.netty4.Netty4H2Writer$class.write(Netty4H2Writer.scala:23)
   at com.twitter.finagle.buoyant.h2.netty4.Netty4H2Writer$$anon$1.write(Netty4H2Writer.scala:74)
   at com.twitter.finagle.buoyant.h2.netty4.Netty4ClientStreamTransport.writeHeaders(Netty4ClientStreamTransport.scala:41)
   at com.twitter.finagle.buoyant.h2.netty4.Netty4ClientDispatcher.apply(Netty4ClientDispatcher.scala:92)
   at com.twitter.finagle.buoyant.h2.netty4.Netty4ClientDispatcher.apply(Netty4ClientDispatcher.scala:23)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.buoyant.h2.LinkerdHeaders$Ctx$Dtab$ClientFilter.apply(LinkerdHeaders.scala:263)
   at com.twitter.finagle.buoyant.h2.LinkerdHeaders$Ctx$Dtab$ClientFilter.apply(LinkerdHeaders.scala:260)
   at com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:79)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.buoyant.h2.LinkerdHeaders$Ctx$Deadline$ClientFilter.apply(LinkerdHeaders.scala:179)
   at com.twitter.finagle.buoyant.h2.LinkerdHeaders$Ctx$Deadline$ClientFilter.apply(LinkerdHeaders.scala:176)
   at com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:79)
   at com.twitter.finagle.ServiceProxy.apply(Service.scala:120)
   at com.twitter.finagle.ServiceProxy.apply(Service.scala:120)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.service.ExceptionRemoteInfoFactory$$anonfun$apply$1$$anon$2.apply(ExceptionRemoteInfoFactory.scala:74)
   at com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:79)
   at com.twitter.finagle.service.FailureAccrualFactory$$anonfun$apply$1$$anon$3.apply(FailureAccrualFactory.scala:339)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.service.StatsFilter.apply(StatsFilter.scala:148)
   at com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:79)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.filter.DtabStatsFilter.apply(DtabStatsFilter.scala:38)
   at com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:79)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.filter.MonitorFilter.apply(MonitorFilter.scala:39)
   at com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:79)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.filter.ExceptionSourceFilter.apply(ExceptionSourceFilter.scala:42)
   at com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:79)
   at com.twitter.finagle.ServiceProxy.apply(Service.scala:120)
   at com.twitter.finagle.service.RefcountedService.apply(RefcountedService.scala:18)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.service.RequeueFilter$$anonfun$com$twitter$finagle$service$RequeueFilter$$applyService$1.apply(RequeueFilter.scala:73)
   at com.twitter.finagle.service.RequeueFilter$$anonfun$com$twitter$finagle$service$RequeueFilter$$applyService$1.apply(RequeueFilter.scala:73)
   at com.twitter.util.Local.let(Local.scala:141)
   at com.twitter.finagle.context.Context$class.let(Context.scala:176)
   at com.twitter.finagle.context.MarshalledContext.let(Context.scala:232)
   at com.twitter.finagle.service.RequeueFilter.com$twitter$finagle$service$RequeueFilter$$applyService(RequeueFilter.scala:72)
   at com.twitter.finagle.service.RequeueFilter.apply(RequeueFilter.scala:111)
   at com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:79)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.filter.ClearContextValueFilter$$anon$1$$anon$2$$anonfun$apply$1.apply(ClearContextValueFilter.scala:26)
   at com.twitter.finagle.filter.ClearContextValueFilter$$anon$1$$anon$2$$anonfun$apply$1.apply(ClearContextValueFilter.scala:26)
   at com.twitter.util.Local.let(Local.scala:141)
   at com.twitter.finagle.context.Context$class.letClear(Context.scala:197)
   at com.twitter.finagle.context.MarshalledContext.letClear(Context.scala:232)
   at com.twitter.finagle.filter.ClearContextValueFilter$$anon$1$$anon$2.apply(ClearContextValueFilter.scala:25)
   at com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:79)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.Failure$ProcessFailures.apply(Failure.scala:308)
   at com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:79)
   at com.twitter.finagle.Service$$anon$1.apply(Service.scala:16)
   at com.twitter.finagle.tracing.TraceInitializerFilter$$anonfun$apply$1.apply(TraceInitializerFilter.scala:55)
   at com.twitter.finagle.tracing.TraceInitializerFilter$$anonfun$apply$1.apply(TraceInitializerFilter.scala:55)
   ...
   ```

**Comment by irachex** [comment]
Another problem is that linkerd seems doesn't handle GOAWAY properly.

According to https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#goaway-frame

> Servers should send GOAWAY before terminating a connection to reliably inform clients which work has been accepted by the server and is being executed.

We add a `keepalive` options for server. server will send GOAWAY then close if client doesn't send any data in `keepalive` seconds. But linkerd report error:

```
E 1019 04:04:33.988 THREAD69: demuxer
java.lang.ClassCastException: io.netty.handler.codec.http2.DefaultHttp2GoAwayFrame cannot be cast to io.netty.handler.codec.http2.Http2StreamFrame
	at com.twitter.finagle.buoyant.h2.netty4.Netty4ClientDispatcher$$anonfun$com$twitter$finagle$buoyant$h2$netty4$Netty4ClientDispatcher$$loop$1$1.apply(Netty4ClientDispatcher.scala:64)
	at com.twitter.util.Future$$anonfun$flatMap$1.apply(Future.scala:1092)
	at com.twitter.util.Future$$anonfun$flatMap$1.apply(Future.scala:1091)
...
```

**Comment by zackangelo** [comment]
Seems like a duplicate of #785.

**Comment by wmorgan** [comment]
@irachex I believe this is fixed in 0.8.4 (per #785). Please feel free to reopen if not.
